pkg/testutil/store.go

```diff
@@ -0,0 +1,125 @@
+package testutil
+
+import (
+	"context"
+	"errors"
+	"io"
+	"testing"
+
+	"github.com/improbable-eng/thanos/pkg/store/storepb"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/tsdb/chunkenc"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// WarningSeries is a series that triggers partial error warning.
+var WarningSeries = storepb.Series{Labels: []storepb.Label{{Name: "WarningSeries"}}}
+
+// StoreClient is test gRPC store API client.
+type StoreClient struct {
+	Values    map[string][]string
+	SeriesSet []storepb.Series
+}
+
+func (s *StoreClient) Info(ctx context.Context, req *storepb.InfoRequest, _ ...grpc.CallOption) (*storepb.InfoResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "not implemented")
+}
+
+func (s *StoreClient) Series(ctx context.Context, req *storepb.SeriesRequest, _ ...grpc.CallOption) (storepb.Store_SeriesClient, error) {
+	return &StoreSeriesClient{ctx: ctx, seriesSet: s.SeriesSet}, nil
+}
+
+func (s *StoreClient) LabelNames(ctx context.Context, req *storepb.LabelNamesRequest, _ ...grpc.CallOption) (*storepb.LabelNamesResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "not implemented")
+}
+
+func (s *StoreClient) LabelValues(ctx context.Context, req *storepb.LabelValuesRequest, _ ...grpc.CallOption) (*storepb.LabelValuesResponse, error) {
+	return &storepb.LabelValuesResponse{Values: s.Values[req.Label]}, nil
+}
+
+// StoreSeriesClient is test gRPC storeAPI series client.
+type StoreSeriesClient struct {
+	// This field just exist to pseudo-implement the unused methods of the interface.
+	storepb.Store_SeriesClient
+	ctx       context.Context
+	i         int
+	seriesSet []storepb.Series
+}
+
+func (c *StoreSeriesClient) Recv() (*storepb.SeriesResponse, error) {
+	if c.i >= len(c.seriesSet) {
+		return nil, io.EOF
+	}
+	s := c.seriesSet[c.i]
+	c.i++
+
+	if s.String() == WarningSeries.String() {
+		// This indicates that we should send warning instead.
+		return storepb.NewWarnSeriesResponse(errors.New("warn")), nil
+
+	}
```

Comment:
Seems better to just give tests direct control over the stream. Sooner or later we'll hit a case where we want to send some specific warning.